src/sb_pointcloud_processing/include/BallExtractorNode.h

```diff
@@ -0,0 +1,134 @@
+#include <string>
+
+class BallExtractorNode {
+  public:
+    /*
+     * @clusters stores the output from DBSCAN
+     */
+    std::vector<pcl::PointCloud<pcl::PointXYZ>> clusters;
+
+    BallExtractorNode(int argc, char** argv, std::string node_name);
+
+    // main entry function
+    void extractBall();
+
+    /*
+     * @clusters: input clusters
+     * @cluster_points: vector to store converted cluster points
+     * @colors: vector to store color for each cluster point
+     * This function takes in @clusters, assigns a different color
+     * for each cluster, and converts each point to geometry_msgs::Point.
+     * It pushes the points and colors to the back of @cluster_points
+     * and @colors.
+     */
+    static void convertClustersToPointsWithColors(
+    std::vector<pcl::PointCloud<pcl::PointXYZ>> clusters,
+    std::vector<geometry_msgs::Point>& cluster_points,
+    std::vector<std_msgs::ColorRGBA>& colors);
+
+  private:
+    ros::Subscriber subscriber;
+    ros::Publisher publisher;
+    ros::Publisher rviz_cluster_publisher;
+    ros::Publisher rviz_ball_publisher;
+
+    /*
+     * @minNeighbours is a hyperparameter to DBSCAN that determines
+     * how many neighbours are required for a point to be considered
+     * a 'core' point
+     */
+    int minNeighbours;
+
+    /*
+     * @radius is a hyperparameter to DBSCAN that determines
+     * how close the points have to be together in order for them to
+     * be considered as 'core' points
+     */
+    float radius;
+
+    /*
+     * scale of the marker points
+     */
+    float scale;
+
+    /*
+     * frame_id for rviz markers
+     */
+    std::string frame_id;
+
+    /*
+     * @pclPtr stores the pointer to the PCL PointCloud after it has
+     * been converted from sensor_msgs PointCloud2
+     */
+    pcl::PointCloud<pcl::PointXYZ>::Ptr pclPtr;
+
+    /*
+     * The callback function is called whenever the node receives a
+     * PointCloud message. It converts sensor_msgs PointCloud2 pointer
+     * to PCL PointCloud pointer and then extracts lines from the PointCloud
+     */
+    void pclCallBack(const sensor_msgs::PointCloud2ConstPtr processed_pcl);
+
+    /*
+     * This function makes a Marker for all points in @clusters
+     * with a different color for each cluster and publishes a message
+     * to rviz for visualization at "~/debug/clusters".
+     */
+    void visualizeClusters();
+
+    /*
+     * This function makes a Marker for the given geometry_msgs::Point
+     * and publishes it to rviz for visualization at "~/debug/ball".
+     */
+    void visualizeBall(geometry_msgs::Point ball);
+
+    /*
+     * Get the center of the cluster
+     * The z value of the center is min and max / 2
+     * The y value of the center is min and max / 2
+     * The x value of the center is the min
+     * @cluster_index: the index of cluster of interest in @clusters
+     */
+    geometry_msgs::Point getCenterOfCluster(unsigned int cluster_index);
```

Comment:
Could we just pass in the cluster by const reference (ie. `const pcl::PointCloud<pcl::PointXYZ>&`)? Seems like it would be less error-prone (can't just pass in a random index and get an "out-of-bounds" error. Then we could also maybe make this `static`?